go/rpcclientlib/obscuro_client.go

```diff
@@ -19,39 +29,228 @@ const (
+
 // Client is used by client applications to interact with the Obscuro node.
 type Client interface {
 	// Call executes the named method via RPC.
 	Call(result interface{}, method string, args ...interface{}) error
 	// Stop closes the client.
 	Stop()
+
+	// SetViewingKey sets the current viewing key on the client to be used for all sensitive request decryption
+	SetViewingKey(viewingKey *ecies.PrivateKey, pubKeyBytes []byte)
+
+	// RegisterViewingKey takes a signature for the public key, it verifies the signed public key matches the currently set private viewing key
+	//	and then submits it to the enclave
+	RegisterViewingKey(signature []byte) error
 }
 
-// A Client implementation that wraps rpc.Client to make calls.
-type clientImpl struct {
-	rpcClient *rpc.Client
+// RPCClient is a Client implementation that wraps rpc.Client to make calls.
+type RPCClient struct {
+	rpcClient        *rpc.Client
+	enclavePublicKey *ecies.PublicKey
+	// todo: add support for multiple keys on the same client?
+	viewingKey     *ecies.PrivateKey // private viewing key to use for decrypting sensitive requests
+	viewingPubKey  []byte            // public viewing key, submitted to the enclave
+	ViewingKeyAddr common.Address    // address that generated the public key
 }
 
-func NewClient(address string) Client {
+func NewClient(address string) *RPCClient {
+	// todo: this is a convenience for testnet but needs to replaced by a parameter and/or retrieved from the target host
+	enclPubECDSA, err := crypto.DecompressPubkey(common.Hex2Bytes(enclavePublicKeyHex))
+	if err != nil {
+		panic(err)
+	}
+	enclavePublicKey := ecies.ImportECDSAPublic(enclPubECDSA)
+
 	rpcClient, err := rpc.Dial(http + address)
 	if err != nil {
 		log.Panic("could not create RPC client on %s. Cause: %s", http+address, err)
 	}
 
-	return &clientImpl{
-		rpcClient: rpcClient,
+	return &RPCClient{
+		rpcClient:        rpcClient,
+		enclavePublicKey: enclavePublicKey,
+	}
+}
+
+// Call handles JSON rpc requests - if the method is sensitive it will encrypt the args before sending the request and
+//		then decrypts the response before returning
+// The result must be a pointer so that package json can unmarshal into it. You
+// 	can also pass nil, in which case the result is ignored.
+func (c *RPCClient) Call(result interface{}, method string, args ...interface{}) error {
+	if !isSensitive(method) {
+		// for non-sensitive methods or when viewing keys are disabled we just delegate directly to the geth RPC client
+		return c.rpcClient.Call(&result, method, args...)
+	}
+
+	// we setup a generic rawResult to receive the response (then we can decrypt it as necessary into the requested result type)
+	var err error
+	var rawResult interface{}
+	if result == nil {
+		// we set result receiver to nil if caller set nil since no return needs to be set
+		rawResult = nil
+	}
+
+	if len(args) == 0 {
```

Comment:
@BedrockSquirrel Can you just encrypt the args no matter what, even if it's an empty list? Think it would make the code much cleaner here.